### PR TITLE
Show last completed at on UI#1078

### PR DESCRIPTION
**Pull request by attilasuto**
#1078

### CODE REVIEW

**Comment by burgerni10** [comment]
Following our internal discussions, a live screen for each south connector will be a good place to display these realtime readonly information (and to separate them from writable settings). These screens will be improved later with other features, like a “web console” to display the connector activity : JF did a poc similar to this, but it’s an improvement for another time.

JF suggestion : create a screen for each protocol (for example: http://localhost:2223/south/myopcua/live) where we will put live informations. This url will be accessible with a button in the list of South (next to the settings button) and a button in the protocol itself (left of the "Points" or "File" button).

**Review by burgerni10 — CHANGES_REQUESTED** [review]
Following our internal discussions, a live screen for each south connector will be a good place to display these realtime readonly information (and to separate them from writable settings). These screens will be improved later with other features, like a “web console” to display the connector activity : JF did a poc similar to this, but it’s an improvement for another time.

JF suggestion : create a screen for each protocol (for example: http://localhost:2223/south/myopcua/live) where we will put live informations. This url will be accessible with a button in the list of South (next to the settings button) and a button in the protocol itself (left of the "Points" or "File" button).

File: src/south/FolderScanner/FolderScanner.class.js [hunk 1]
```diff
@@ -25,6 +25,10 @@ class FolderScanner extends ProtocolHandler {
     this.minAge = minAge
     this.regex = new RegExp(regex)
     this.compression = compression
+
+    this.lastOnScanAt = null
+    this.lastAddFileAt = null
```
Comment (jfhenon) [inline src/south/FolderScanner/FolderScanner.class.js:30]: this.lastOnScanAt = null
    this.lastAddFileAt = null
could be part of the ProtocolHandlerClass

File: src/south/FolderScanner/FolderScanner.class.js [hunk 2]
```diff
@@ -34,6 +38,7 @@ class FolderScanner extends ProtocolHandler {
    */
   async onScan(scanMode) {
     this.logger.debug(`FolderScanner activated on scanMode: ${scanMode}.`)
+    this.lastOnScanAt = new Date().getTime()
```
Comment (jfhenon) [inline src/south/FolderScanner/FolderScanner.class.js:41]: This variable can be maintained by the parent class ProtocolHandler (before calling onScan)

File: src/south/FolderScanner/FolderScanner.class.js [hunk 3]
```diff
@@ -126,6 +131,18 @@ class FolderScanner extends ProtocolHandler {
       await this.setConfig(filename, `${stats.mtimeMs}`)
     }
   }
+
+  /**
```
Comment (jfhenon) [inline src/south/FolderScanner/FolderScanner.class.js:135]: Nothing special for this class. The default method could be used.

File: src/south/ProtocolHandler.class.js [hunk 4]
```diff
@@ -174,6 +176,16 @@ class ProtocolHandler {
       setTimeout(resolve, timeout)
     })
   }
+
+  /**
+   * Get live status.
+   * @returns {object} - The live status
+   */
+  getStatus() {
```
Comment (jfhenon) [inline src/south/ProtocolHandler.class.js:184]: The default implementation should not be surcharged and could return standard variables. 

if we need a specific variable for this handler:

getStatus() {
const status = super.getStatus()
status.anykey = anyvalue
return status
}

**Comment by jfhenon** [comment]
@attilasuto 
image
I would either put the (i) close to the name UA or I would turn it into a normal button close to the points button.
@burgerni10 your view?

File: src/south/ProtocolHandler.class.js [hunk 5]
```diff
@@ -174,6 +180,19 @@ class ProtocolHandler {
...
+    const status = { 'Last scan time': this.lastOnScanAt ? new Date(this.lastOnScanAt).toLocaleString() : 'Never' }
+    if (this.handlesFiles) {
+      status['Last file added time'] = this.lastAddFileAt ? new Date(this.lastAddFileAt).toLocaleString() : 'Never'
+      status['Number of files added'] = this.addFileCount
+    }
```
Comment (jfhenon) [inline src/south/ProtocolHandler.class.js:193]: need to also track Last value and number of values, last timestamp, last pointid,

**Review by jfhenon — CHANGES_REQUESTED** [review]
Made some comment on UI 
Also on default status for values

**Comment by burgerni10** [comment]
> @attilasuto
> image
> I would either put the (i) close to the name UA or I would turn it into a normal button close to the points button.
> @burgerni10 your view?

I am not sure about this.

The way it is now, it will always display the (i) at the same place, making UI more consistent. Next to the name, it would change a little bit regarding the length of the breadcrumb.

But I agree with the second part of your message : it should look like a normal button. The way it is designed now make it look like the help button (?) or something where a popover appears once we click on it. @attilasuto, what do you think ? 

I suggest a button with "Status" in it.

File: src/south/ProtocolHandler.class.js (hunk 5 above)
Comment (kukukk) [reply src/south/ProtocolHandler.class.js:193]: I was planning to include point related informations when we implement it for the first point compatible South.

**Comment by attilasuto** [comment]
> > @attilasuto
> > image
> > I would either put the (i) close to the name UA or I would turn it into a normal button close to the points button.
> > @burgerni10 your view?
> 
> I am not sure about this.
> 
> The way it is now, it will always display the (i) at the same place, making UI more consistent. Next to the name, it would change a little bit regarding the length of the breadcrumb.
> 
> But I agree with the second part of your message : it should look like a normal button. The way it is designed now make it look like the help button (?) or something where a popover appears once we click on it. @attilasuto, what do you think ?
> 
> I suggest a button with "Status" in it.

I agree it would fit more better to have a button with "Status" on the settings screen and keep the (i) on the south listing screen. I will change this as required.

File: src/south/ProtocolHandler.class.js (hunk 5 above)
Comment (jfhenon) [reply src/south/ProtocolHandler.class.js:193]: I think the default status could be added now as well.

**Review by jfhenon — COMMENTED** [review]
@kukukk minor comments on ProtocolHandler
line 44 -> this.compressionLevel should this be an engine parameter?
line 66 -> maybe we could avoid to ask the south to use super.onScan by asking the South handler to define onScanImplementation that would be call by OnScan.

**Review by jfhenon — APPROVED** [review]

**Comment by burgerni10** [comment]
I approve the Pr, but I suggest some changes in issue #1173  
I put these changes in a separate issue because it's more general than just the "Status" button

**Review by burgerni10 — APPROVED** [review]

**Pull request merged**